stream/src/main/java/org/apache/kafka/streaming/processor/internals/RecordQueue.java

```diff
@@ -17,22 +17,24 @@
 
 package org.apache.kafka.streaming.processor.internals;
 
-import org.apache.kafka.clients.consumer.ConsumerRecord;
 import org.apache.kafka.common.TopicPartition;
 
 import java.util.ArrayDeque;
+import java.util.PriorityQueue;
 
 /**
- * RecordQueue is a queue of {@link StampedRecord} (ConsumerRecord + timestamp).
+ * RecordQueue is a FIFO queue of {@link StampedRecord} (ConsumerRecord + timestamp). It also keeps track of the
+ * partition timestamp defined as the minimum timestamp of records in its queue; in addition, its partition
+ * timestamp is monotonically increasing such that once it is advanced, it will not be decremented.
  */
 public class RecordQueue {
 
     private final SourceNode source;
     private final TopicPartition partition;
-    private final ArrayDeque<StampedRecord> queue = new ArrayDeque<>();
-    private final TimestampTracker<ConsumerRecord<Object, Object>> timestampTracker = new MinTimestampTracker<>();
+    private final ArrayDeque<StampedRecord> fifoQueue = new ArrayDeque<>();
+    private final PriorityQueue<StampedRecord> timeQueue = new PriorityQueue<>();
```

Comment:
@guozhangwang 
Why did you remove MinTimestampTracker? PriorityQueue is not a good data structure for tracking the min value in a sliding window. The insertion costs O(log N), The removal costs O(N + log N) MinTimestampTracker's amortized costs of insertion and deletion are O(1).

Please resurrect MinTimestampTracker. You can rename it if you want.